pkg/services/searchV2/index.go

```diff
@@ -0,0 +1,357 @@
+package searchV2
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/searchV2/extract"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/services/store"
+)
+
+type dashboardLoader interface {
+	LoadDashboards(ctx context.Context, orgID int64, dashboardUID string) ([]dashboard, error)
+}
+
+type DashboardIndex struct {
+	mu         sync.RWMutex
+	loader     dashboardLoader
+	dashboards map[int64][]dashboard
+	eventStore store.EntityEventsService
+	logger     log.Logger
+}
+
+type dashboard struct {
+	id       int64
+	uid      string
+	isFolder bool
+	folderID int64
+	slug     string
+	created  time.Time
+	updated  time.Time
+	info     *extract.DashboardInfo
+}
+
+func NewDashboardIndex(eventStore store.EntityEventsService, loader dashboardLoader) *DashboardIndex {
+	return &DashboardIndex{
+		loader:     loader,
+		dashboards: map[int64][]dashboard{},
+		eventStore: eventStore,
+		logger:     log.New("dashboardIndex"),
+	}
+}
+
+func (i *DashboardIndex) run(ctx context.Context) error {
+	fullUpdateTicker := time.NewTicker(5 * time.Minute)
+	defer fullUpdateTicker.Stop()
+
+	partialUpdateTicker := time.NewTicker(5 * time.Second)
+	defer partialUpdateTicker.Stop()
+
+	var lastEventID int64
+	lastEvent, err := i.eventStore.GetLastEvent(context.Background())
+	if err != nil {
+		return err
+	}
+	if lastEvent != nil {
+		lastEventID = lastEvent.Id
+	}
+
+	// Build on start for orgID 1 but keep lazy for others.
+	_, err = i.getDashboards(ctx, 1)
+	if err != nil {
+		return fmt.Errorf("can't build dashboard search index for org ID 1: %w", err)
+	}
+
+	for {
+		select {
+		case <-partialUpdateTicker.C:
+			lastEventID = i.loadIndexUpdates(context.Background(), lastEventID)
+		case <-fullUpdateTicker.C:
+			i.reIndexFromScratch()
```

Comment:
A bit worrying about this. I don't know how this can affect huge installations. Any idea what we can do here? I think  adding time metric (histogram) to this right from the start can be helpful?

Reply:
I think it would be worth doing at least these:

1. allowing queries to go through while we build the new index, ie. executing `i.mu.Lock()` only after we have the new index ready to go
2. adding an integration test with large number of dashboards (5000?) to make sure we don't have any obvious cpu bottlenecks
3. paginate the dashboard query (https://github.com/grafana/grafana/blob/5f8e2784a97b45a2431ba2e2b4e0221059ff9722/pkg/services/searchV2/index.go#L255) - with 1000 dashboards and 10kb per dash we are currently expecting a 10mb json payload
4. deploy it on HG, first `dev` and then ops.grafana.net to see how it behaves :)

Reply:
Implemented 1 and 3

Reply:
> deploy it on HG, first dev and then ops.grafana.net to see how it behaves :) We need to remove requiresDevMode before testing the feature on HG

<3

Reply:
Make sure to log the time taken to index the data. both full and partial. Then we can gather stats about it on HG using Loki using `duration_seconds(label_identifier)`

https://grafana.com/docs/loki/latest/logql/metric_queries/#unwrapped-range-aggregations

Reply:
@bergquist added several log lines:

```go
i.logger.Info("Re-indexed dashboards for organization", "orgId", orgID, "org_re_index_elapsed", time.Since(started))
...
i.logger.Info("Full re-indexing finished", "full_re_index_elapsed", time.Since(started))
...
i.logger.Info("Index updates applied", "index_events_applied_elapsed", time.Since(started), "numEvents", len(events))
```

Is this sufficient to look in Loki? Info level and unique labels right?

Reply:
@FZambia looks good! :)